src/libraries/System.Private.CoreLib/src/System/Reflection/ModifiedType.cs

```diff
@@ -0,0 +1,192 @@
+                modifiedType = new ModifiedContainerType(unmodifiedType, root, nestedSignatureIndex, nestedSignatureParameterIndex);
+            }
+            else if (unmodifiedType.IsGenericType)
+            {
+                modifiedType = new ModifiedGenericType(unmodifiedType, root, nestedSignatureIndex, nestedSignatureParameterIndex);
+            }
+            else
+            {
+                modifiedType = new ModifiedStandaloneType(unmodifiedType, root, nestedSignatureIndex, nestedSignatureParameterIndex);
+            }
+
+            return modifiedType;
+        }
+
+        /// <summary>
+        /// The root signature's parameter index (0 for properties, 1 for fields, 0..n for methods).
+        /// A value of -1 means the value is not used because the node is not a root.
+        /// </summary>
+        protected int RootSignatureParameterIndex => Root._rootSignatureParameterIndex;
+
+        /// <summary>
+        /// The nested signature's index into the recursive type tree.
+        /// A signature exists for function pointers and generic types.
+        /// </summary>
+        // For delegate*<void>[]: 0 for the function pointer since nested by the array
+        // For delegate*<delegate*<void>>: -1 for the outer (since not nested); 0 for the inner
+        // For delegate*<delegate*<void>>[]: 0 for the outer; 1 for the inner
+        protected int NestedSignatureIndex => _nestedSignatureIndex;
+
+        /// <summary>
+        /// From a given signature from <see cref="NestedSignatureIndex"/>, which parameter index does
+        /// the node belong to. 0 for return; 1..n for parameters.
+        /// A value of -1 means the value is not used because the node is not a signature parameter.
+        /// /// </summary>
+        protected int NestedSignatureParameterIndex => _nestedSignatureParameterIndex;
+
+        /// <summary>
+        /// The root node which contains the signature reference and _rootSignatureParameterIndex.
+        /// </summary>
+        public ModifiedType Root
+        {
+            get
+            {
+                Debug.Assert( _root != null );
+                return _root;
+            }
+        }
+
+        public override Type[] GetRequiredCustomModifiers()
+        {
+            // No caching is performed; as is the case with FieldInfo.GetCustomModifiers etc.
+            return GetCustomModifiers(required: true);
+        }
+
+        public override Type[] GetOptionalCustomModifiers()
+        {
+            // No caching is performed; as is the case with FieldInfo.GetCustomModifiers etc.
+            return GetCustomModifiers(required: false);
+        }
+
+        private Type[] GetCustomModifiers(bool required)
+        {
+            Type[] modifiers = EmptyTypes;
+
+            if (_nestedSignatureParameterIndex >= 0)
+            {
+                modifiers = GetCustomModifiersFromSignature(required);
+            }
+            else if (ReferenceEquals(this, Root))
```

Comment:
Can we just call `GetCustomModifiersFromSignature` for the top-level modifiers as well? I do not see what we are saving by trying to get them from the owning entity.

Reply:
I'll update to do that since it makes it simpler for now. However once Mono\NativeAot adds support, this area may change again (intended to be shared code)